File: backend/app/Models/Payment.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\MorphTo;

#[Fillable([
    'user_id', 'payable_type', 'payable_id', 'provider', 'amount', 'currency',
    'reference', 'payer_phone', 'status', 'paid_at', 'meta',
])]
class Payment extends Model
{
    public const STATUS_PENDING = 'pending';

    public const STATUS_PAID = 'paid';

    public const STATUS_FAILED = 'failed';

    public const STATUS_CANCELLED = 'cancelled';

    public const STATUS_REFUNDED = 'refunded';

    public const PROVIDER_WAVE = 'wave';

    public const PROVIDER_ORANGE_MONEY = 'orange_money';

    protected function casts(): array
    {
        return [
            'amount' => 'integer',
            'paid_at' => 'datetime',
            'meta' => 'array',
        ];
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function payable(): MorphTo
    {
        return $this->morphTo();
    }

    public function transactions(): HasMany
    {
        return $this->hasMany(PaymentTransaction::class);
    }

    public function isPaid(): bool
    {
        return $this->status === self::STATUS_PAID;
    }
}

File: lib/models/payment.dart
class Payment {
  const Payment({
    required this.id,
    required this.reference,
    required this.provider,
    required this.amount,
    required this.currency,
    required this.status,
    required this.payableType,
    this.checkoutUrl,
    this.paidAt,
    required this.createdAt,
  });

  final int id;
  final String reference;
  final String provider;
  final int amount;
  final String currency;
  final String status;
  final String payableType;
  final String? checkoutUrl;
  final DateTime? paidAt;
  final DateTime createdAt;

  bool get isPaid => status == 'paid';
  bool get isPending => status == 'pending';

  factory Payment.fromJson(Map<String, dynamic> json) => Payment(
    id: json['id'] as int,
    reference: json['reference'] as String,
    provider: json['provider'] as String,
    amount: json['amount'] as int,
    currency: json['currency'] as String,
    status: json['status'] as String,
    payableType: json['payable_type'] as String,
    checkoutUrl: json['checkout_url'] as String?,
    paidAt: json['paid_at'] != null
        ? DateTime.tryParse(json['paid_at'] as String)
        : null,
    createdAt: DateTime.parse(json['created_at'] as String),
  );
}
